### DISCUSSION TITLE
Use `uv` for managing dependencies / local development

### DISCUSSION THREAD

**Discussion by samuelhwilliams** [communitiesuk/funding-service-requests-for-comments#2] [General]
Title: Use `uv` for managing dependencies / local development
    Owner: samuel.williams@communities.gov.uk
    Collaborator(s):
    Created on: 2024-10-31
    Status: Draft
    Finalised on: TBC

# Overview

This proposal discusses moving from a few different tools for managing our Python ecosystem (mostly locally) to a single tool, `uv`.

# What is the current state?

Currently, we use a number of tools to manage dependencies and local development:

* `pip` for installing dependencies
* `pip-tools` for compiling/locking dependencies
* `pyenv` for managing python versions

`pip` is fairly slow at installing dependencies and does not have the most friendly developer experience. `uv` is a modern tool that aims to improve developer experience and run fast. It has functionality to replace all three of these tools in one go.

# Why should we change?

* Make it easier to add/remove/edit dependencies.
* Have a faster, more streamlined developer experience - including onboarding/setup.
* Stay up to date with the Python ecosystem, eg [PEP 735](https://peps.python.org/pep-0735/) which adds the concept of "dependency groups".

Managing dependencies locally is a somewhat slow and annoying process, due to:

* `pip`'s CLI not having very user-friendly options
* `pip`'s dependency resolver being slow and emitting less friendly error messages on failure
* Needing to manage virtual environments manually (creating them, keeping them in sync, etc)

# How should we address this?

`uv` is an actively-maintained tool managed by [astral.sh](https://astral.sh/), which has been working hard to improve the Python ecosystem. We use one of their other tools, `ruff`, to manage linting and formatting of parts of the Funding Service (mainly post award - but likely to be everything soon).

`uv` runs so quickly that it will generally run an automatic create+sync before any command you run, so you don't need to manually do nearly as much to manage your local environments. This goes for installing the correct pre-compiled python binary for your application as well.

`uv` uses `pyproject.toml` to store top-level dependencies, and manages both core and dev dependencies in a PEP-compliant format there. It supports extras (eg for utils) and dependency groups (eg for dev dependencies) consistently.

| Task | old command | new command |
| --- | --- | --- |
| Setting up for the first time | `pyenv install 3.10 && python3.10 -m venv .venv && source .venv/bin/activate && pip install -r requirements-dev.txt`| `uv sync` |
| Compiling/locking the dependency tree | `pip-compile requirements.in && pip-compile requirements-dev.in` |uv lock  |
| Adding/removing a dependency | manually edit `requirements[-dev].in` && pip-compile requirements.in && pip-compile requirements-dev.in | `uv add/remove --dev <package>` |
| Updating a package | manually edit `requirements[-dev].in` && pip-compile requirements.in && pip-compile requirements-dev.in | `uv lock -P <package>` |
| Installing a python version | `pyenv install 3.10` | `uv python install 3.10` (but you shouldn't need to manually do this) |

We will need to do migrate the existing `.in` files into `pyproject.toml`. I prepared a couple of commands that can do this for a given repo by setting `requirements.txt` as a constraints file, and then doing `uv add --requirements requirements.in`. And the same for `requirements-dev.in`.

# What alternatives are we discarding?

There are a number of ways of managing dependencies for python projects:

* `pip-tools` - stay as we are
* [`poetry`](https://python-poetry.org/)
* [`pipenv`](https://pipenv.pypa.io/en/latest/)
* [`pdm`](https://pdm-project.org/en/latest/)

These all address the "dependencies" part of the process, but only `pdm` also manages python installation locally. Poetry and pipenv use slower dependency resolution that makes managing dependencies more annoying than it needs to be.

# Who will be affected?

This will affect all developers, who will need to switch over to and learn how to use `uv` and what the equivalents are for any commands they use locally.

# What are the key risks to manage or mitigate?

The risks with switching out a tool here are minimal. We can maintain backwards compatibility with pip by exporting `uv` lockfile to a standard `requirements.txt` file, which pip/etc could still use directly.

## Dependabot

Dependabot does not currently know how to update `uv.lock` files. [Renovate](https://docs.renovatebot.com/) does. We could consider switching to that. Or we could use `uv export requirements.txt` to export `uv.lock` to a standard requirements file that Dependabot understands, and then have actions that run on Dependabot PRs which manage updating our uv lock file.

## Paketo

Paketo does not understand `uv.lock` files. We can keep a `requirements.txt` using `uv export`, and then nothing would need to change with Paketo or our deployment process. Alternatively, we could write a custom Paketo buildpack that runs the `uv sync` command (probably not recommended).

**Comment by samuelhwilliams** [comment]
P.S. - if you want to see what this looks like with the existing local setup, please checkout the `proto/uv-requirements` branch on docker runner and on each of the apps within that.

Assuming you're on latest clean main for each repo, from the docker runner repo, you can run:

```bash
git checkout proto/uv-requirements

cd apps

for d in $(ls); do cd ${d}; echo ${d}; git checkout proto/uv-requirements; cd ..; done

cd ..
```

**Reply by albertkol** [reply]
You mean `git checkout`? :eyes:

**Reply by samuelhwilliams** [reply]
oops 🤦  thanks

**Comment by albertkol** [comment]
I will give it a try soon. See how much quicker it is.

**Comment by srh-sloan** [comment]
I like this idea. Bit concerned about the paketo issue though and I don't like the idea of storing a `requirements.txt` file anyway just to get round that. Could we run `uv export` as part of the pipeline to just generate that paketo needs at that point?

**Reply by samuelhwilliams** [reply]
> Could we run uv export as part of the pipeline to just generate that paketo needs at that point?

That feels smart and I hadn't thought of that. I don't see immediately why that wouldn't work - will try to check and confirm it.

**Reply by samuelhwilliams** [reply]
Workflows change: https://github.com/communitiesuk/funding-service-design-workflows/compare/proto/uv-requirements
Example app change: https://github.com/communitiesuk/funding-service-design-post-award-data-store/commit/d36128de9c175d04bfb671c5bb4666ad1a65d081

Example paketo run: https://github.com/communitiesuk/funding-service-design-post-award-data-store/actions/runs/11687360487/job/32545283128

Looks like a good idea ✅

**Reply by samuelhwilliams** [reply]
Side bonus/proof of `uv` speed: installing dependencies and building the static assets for that repo is about ~25 seconds with pip, and ~3s with uv.

## pip
<img width="893" alt="image" src="https://github.com/user-attachments/assets/fe0c643b-02af-4c00-936c-4cb7469f40c8">

## uv
<img width="1019" alt="image" src="https://github.com/user-attachments/assets/7550c70d-47fd-4e14-8fbe-155502806a19">

**Comment by sfount** [comment]
Looks sensible to me! It would be good to have confidence that we can have dependabot continue to prompt for upgrades (using the mentioned requirements backwards compatibility) and then either tell developers how to resolve those in bulk or run a GitHub Action as also mentioned -- I like the suggestion of trying that out on a repository that isn't yet going through to production.

Appreciate that the existing dependabot process needs a lot of work and many dependencies are out of date and manually updated, hopefully moving to less repositories to manage will also help this run smoother.

**Reply by albertkol** [reply]
How else will we be able to ignore those dependabot emails, if we don't get them anymore?!

**Reply by samuelhwilliams** [reply]
@gidsg and I made a little toy repo with renovate on it. Here are the PRs:

* https://github.com/communitiesuk/funding-service-renovate-test/pull/1
* https://github.com/communitiesuk/funding-service-renovate-test/pull/3
* https://github.com/communitiesuk/funding-service-renovate-test/pull/6

I think we should be OK to use this until Dependabot catches up with `uv.lock`, and then switch back to dependabot (since that's maybe more common for MHCLG)?
